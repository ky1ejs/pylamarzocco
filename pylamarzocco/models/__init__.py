"""Models for the La Marzocco API."""

from ._authentication import AccessToken, RefreshTokenRequest, SigninTokenRequest
from ._config import (
    BackFlush,
    BaseDoseSettings,
    BrewByWeightDoses,
    BrewByWeightDoseSettings,
    CoffeeBoiler,
    DosePulsesType,
    DoseSettings,
    GroupDosesSettings,
    HotWaterDose,
    InOutTime,
    MachineStatus,
    NextStatus,
    PreBrewInfusionTime,
    PrebrewInfusionTimeLists,
    PreBrewing,
    PrebrewSettingTimes,
    PreExtraction,
    PreExtractionBase,
    PreExtractionBaseTimes,
    PreExtractionInOutTimes,
    PreExtractionPreBrewInfusionTimes,
    SecondsInOut,
    SteamBoilerLevel,
    SteamBoilerTemperature,
    ThingConfig,
    ThingDashboardConfig,
    ThingDashboardWebsocketConfig,
    ThingScale,
    ThingSettings,
)
from ._general import (
    BaseWidget,
    BaseWidgetOutput,
    CommandResponse,
    Thing,
    WebSocketDetails,
    Widget,
)
from ._schedule import (
    SchedulingSettings,
    SmartWakeUpScheduleWebsocketConfig,
    SmartWakeUpSleepSettings,
    WakeUpScheduleSettings,
)
from ._statistics import (
    CoffeeAndFlushTrend,
    CoffeeHistoryEvent,
    LastCoffee,
    LastCoffeeList,
    ThingStatistics,
)
from ._update import FirmwareSettings, FirmwareVersion, UpdateDetails

__all__ = [
    "AccessToken",
    "BackFlush",
    "BaseDoseSettings",
    "BaseWidget",
    "BaseWidgetOutput",
    "BrewByWeightDoseSettings",
    "BrewByWeightDoses",
    "CoffeeAndFlushTrend",
    "CoffeeBoiler",
    "CoffeeHistoryEvent",
    "CommandResponse",
    "ThingConfig",
    "ThingDashboardConfig",
    "ThingDashboardWebsocketConfig",
    "DosePulsesType",
    "DoseSettings",
    "FirmwareSettings",
    "FirmwareVersion",
    "GroupDosesSettings",
    "HotWaterDose",
    "InOutTime",
    "LastCoffee",
    "LastCoffeeList",
    "MachineStatus",
    "NextStatus",
    "PreBrewInfusionTime",
    "PreBrewing",
    "PreExtraction",
    "PreExtractionBase",
    "PreExtractionBaseTimes",
    "PreExtractionInOutTimes",
    "PreExtractionPreBrewInfusionTimes",
    "PrebrewInfusionTimeLists",
    "PrebrewSettingTimes",
    "RefreshTokenRequest",
    "SchedulingSettings",
    "SecondsInOut",
    "SigninTokenRequest",
    "SmartWakeUpScheduleWebsocketConfig",
    "SmartWakeUpSleepSettings",
    "SteamBoilerLevel",
    "SteamBoilerTemperature",
    "Thing",
    "ThingScale",
    "ThingSettings",
    "ThingStatistics",
    "UpdateDetails",
    "WakeUpScheduleSettings",
    "WebSocketDetails",
    "Widget",
]
