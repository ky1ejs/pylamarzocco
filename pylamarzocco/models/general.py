"""Models for general info"""

from dataclasses import dataclass, field

from mashumaro import field_options
from mashumaro.mixins.json import DataClassJSONMixin


@dataclass(kw_only=True)
class CommandResponse(DataClassJSONMixin):
    """Response for change setting endpoint"""
    id: str
    status: str
    error_code: str | None = field(
        metadata=field_options(alias="errorCode")
    )