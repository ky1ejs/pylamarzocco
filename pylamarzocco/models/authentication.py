"""Models for authentication."""

from time import time
from dataclasses import dataclass, field

from mashumaro import field_options
from mashumaro.mixins.json import DataClassJSONMixin

from pylamarzocco.const import TOKEN_EXPIRATION


@dataclass(kw_only=True)
class AccessToken(DataClassJSONMixin):
    """Token response model."""

    id: str 
    access_token: str = field(metadata=field_options(alias="accessToken"))
    refresh_token: str = field(metadata=field_options(alias="refreshToken"))
    token_type: str = field(metadata=field_options(alias="tokenType"))
    username: str
    email: str
    expires_in: float = field(default=time() + TOKEN_EXPIRATION)


@dataclass(kw_only=True)
class TokenRequest(DataClassJSONMixin):
    """Token request model."""

    username: str
    password: str