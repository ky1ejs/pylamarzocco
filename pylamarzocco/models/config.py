"""Models for device configuration."""

from __future__ import annotations
from typing import Annotated

from dataclasses import dataclass, field
from datetime import datetime, timezone

from mashumaro import field_options
from mashumaro.mixins.json import DataClassJSONMixin
from mashumaro.types import Discriminator

from pylamarzocco.models.general import CommandResponse

from pylamarzocco.const import MachineState, PreExtractionMode

@dataclass(kw_only=True)
class DeviceConfig(DataClassJSONMixin):
    """Device configuration."""

    connected: bool
    removed_widgets: list[BaseWidget] = field(
        metadata=field_options(alias="removedWidgets"), default_factory=list
    )
    connection_date: int = field(metadata=field_options(alias="connectionDate"))
    widgets: list[Widget] = field(default_factory=list)
    uuid: str
    commands: list[CommandResponse]


@dataclass(kw_only=True)
class BaseWidget(DataClassJSONMixin):
    """Base widget configuration."""

    code: str
    index: int


@dataclass(kw_only=True)
class Widget(BaseWidget):
    """Widget configuration."""

    output: list[
        Annotated[WidgetBaseOutput, Discriminator(field="code", include_subtypes=True)]
    ]

@dataclass(kw_only=True)
class WidgetBaseOutput(DataClassJSONMixin):
    """Widget configuration."""

@dataclass(kw_only=True)
class CMMachineStatus(WidgetBaseOutput):
    """Machine status configuration."""

    status: MachineState
    available_modes: list[MachineState] = field(
        metadata=field_options(alias="availableModes")
    )
    mode: MachineState
    next_state: MachineState | None = field(metadata=field_options(alias="nextState"))
    brewing_start_time: datetime | None = field(
        metadata=field_options(
            alias="brewingStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )


@dataclass(kw_only=True)
class CMCoffeeBoiler(WidgetBaseOutput):
    """Coffee boiler configuration."""

    status: MachineState
    enabled: bool
    enabled_supported: bool = field(metadata=field_options(alias="enabledSupported"))
    target_temperature: float = field(metadata=field_options(alias="targetTemperature"))
    target_temperature_min: int = field(
        metadata=field_options(alias="targetTemperatureMin")
    )
    target_temperature_max: int = field(
        metadata=field_options(alias="targetTemperatureMax")
    )
    target_temperature_step: float = field(
        metadata=field_options(alias="targetTemperatureStep")
    )
    ready_start_time: datetime | None = field(
        metadata=field_options(
            alias="readyStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )

@dataclass(kw_only=True)
class CMSteamBoilerLevel(WidgetBaseOutput):
    """Steam boiler level configuration."""
    status: MachineState # TODO: correct type
    enabled: bool
    enabled_supported: bool = field(metadata=field_options(alias="enabledSupported"))
    target_level: float = field(metadata=field_options(alias="targetLevel"))
    target_level_supported: bool = field(
        metadata=field_options(alias="targetLevelSupported")
    )
    ready_start_time: datetime | None = field(
        metadata=field_options(
            alias="readyStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )

@dataclass(kw_only=True)
class PreExtractionBase(WidgetBaseOutput):
    """Pre-extraction configuration."""
    available_modes: list[PreExtractionMode] = field(
        metadata=field_options(alias="availableModes")
    )
    mode: PreExtractionMode

@dataclass(kw_only=True)
class CMPreExtraction(PreExtractionBase):
    """Pre-extraction configuration."""

    times: InOutTime

@dataclass(kw_only=True)
class CMPreBrewing(PreExtractionBase):
    """Pre-brewing configuration."""

    times: PrebrewInfusionTimeLists
    dose_index_supported: bool = field(metadata=field_options(alias="doseIndexSupported"), default=False)

@dataclass(kw_only=True)
class InOutTime(DataClassJSONMixin):
    """In and out time configuration."""

    time_in: PreExtractionInOutTimes = field(metadata=field_options(alias="In"))
    time_out: PreExtractionInOutTimes = field(metadata=field_options(alias="Out"))

@dataclass(kw_only=True)
class SecondsInOut(DataClassJSONMixin):
    """Seconds in and out configuration."""
    seconds_in: float = field(metadata=field_options(alias="In"))
    seconds_out: float = field(metadata=field_options(alias="Out"))

@dataclass(kw_only=True)
class PreBrewInfusionTime(DataClassJSONMixin):
    """Prebrew/-infusion configuration."""
    pre_infusion: float = field(metadata=field_options(alias="PreInfusion"))
    pre_brewing: float = field(metadata=field_options(alias="PreBrewing"))

@dataclass(kw_only=True)
class PreExtractionBaseTimes[T](DataClassJSONMixin):
    """Pre-extraction times configuration."""
    seconds_min: T = field(
        metadata=field_options(alias="secondsMin")
    )
    seconds_max: T = field(
        metadata=field_options(alias="secondsMax")
    )
    seconds_step: T = field(
        metadata=field_options(alias="secondsStep")
    )   


@dataclass(kw_only=True)
class PreExtractionPreBrewInfusionTimes(PreExtractionBaseTimes[PreBrewInfusionTime]):
    """Pre-extraction times configuration."""
    seconds: float

@dataclass(kw_only=True)
class PreExtractionInOutTimes(PreExtractionBaseTimes[SecondsInOut]):
    """Pre-extraction times configuration."""
    dose_index: str = field(metadata=field_options(alias="doseIndex"))


@dataclass(kw_only=True)
class PrebrewInfusionTimeLists(DataClassJSONMixin):
    """Prebrew/-infusion configuration."""
    pre_infusion: list[PreExtractionInOutTimes] = field(metadata=field_options(alias="PreInfusion"))
    pre_brewing: list[PreExtractionInOutTimes] = field(metadata=field_options(alias="PreBrewing"))

@dataclass(kw_only=True)
class CMBackFlush(WidgetBaseOutput):
    """Backflush configuration."""

    enabled: bool
    last_cleaning_start_time: datetime | None = field(
        metadata=field_options(
            alias="lastCleaningStartTime",
            serialize=lambda ts: datetime.fromtimestamp(ts, timezone.utc),
        ),
        default=None,
    )