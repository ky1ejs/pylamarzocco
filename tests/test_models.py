"""Testing models serialization and deserialization."""

from syrupy import SnapshotAssertion
from pylamarzocco.models.config import DeviceConfig

from .conftest import load_fixture

async def test_device_config(snapshot: SnapshotAssertion) -> None:
    """Test the config model serialization."""

    fixture = load_fixture("machine", "config_micra.json")
    device = DeviceConfig.from_dict(fixture)
    assert device.to_dict() == snapshot