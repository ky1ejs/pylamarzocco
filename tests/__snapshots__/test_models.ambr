# serializer version: 1
# name: test_device_config
  dict({
    'commands': list([
      dict({
        'error_code': None,
        'id': '14939275-fdc3-4aaf-85be-1a1a7d418188',
        'status': 'Success',
      }),
    ]),
    'connected': True,
    'connection_date': 1742439620235,
    'removed_widgets': list([
      dict({
        'code': 'CMMachineGroupStatus',
        'index': 1,
      }),
      dict({
        'code': 'CMSteamBoilerTemperature',
        'index': 1,
      }),
      dict({
        'code': 'CMGroupDoses',
        'index': 1,
      }),
      dict({
        'code': 'CMPreInfusionEnable',
        'index': 1,
      }),
      dict({
        'code': 'CMPreInfusion',
        'index': 1,
      }),
      dict({
        'code': 'CMBrewByWeightDoses',
        'index': 1,
      }),
      dict({
        'code': 'CMCupWarmer',
        'index': 1,
      }),
      dict({
        'code': 'CMHotWaterDose',
        'index': 1,
      }),
      dict({
        'code': 'CMAutoFlush',
        'index': 1,
      }),
      dict({
        'code': 'CMRinseFlush',
        'index': 1,
      }),
      dict({
        'code': 'CMSteamFlush',
        'index': 1,
      }),
      dict({
        'code': 'CMNoWater',
        'index': 1,
      }),
      dict({
        'code': 'GMachineStatus',
        'index': 1,
      }),
      dict({
        'code': 'GDoses',
        'index': 1,
      }),
      dict({
        'code': 'GSingleDoseMode',
        'index': 1,
      }),
      dict({
        'code': 'GBaristaLight',
        'index': 1,
      }),
      dict({
        'code': 'GHopperOpened',
        'index': 1,
      }),
      dict({
        'code': 'GMirrorDoses',
        'index': 1,
      }),
      dict({
        'code': 'GMoreDose',
        'index': 1,
      }),
      dict({
        'code': 'GGrindWith',
        'index': 1,
      }),
    ]),
    'uuid': '227e486d-583d-4e69-8113-2e89ac0bce53',
    'widgets': list([
      dict({
        'code': 'CMMachineStatus',
        'index': 1,
        'output': dict({
          'available_modes': list([
            'BrewingMode',
            'StandBy',
          ]),
          'brewing_start_time': None,
          'mode': 'StandBy',
          'next_status': None,
          'status': 'StandBy',
        }),
      }),
      dict({
        'code': 'CMCoffeeBoiler',
        'index': 1,
        'output': dict({
          'enabled': True,
          'enabled_supported': False,
          'ready_start_time': None,
          'status': 'StandBy',
          'target_temperature': 94.0,
          'target_temperature_max': 100,
          'target_temperature_min': 80,
          'target_temperature_step': 0.1,
        }),
      }),
      dict({
        'code': 'CMSteamBoilerLevel',
        'index': 1,
        'output': dict({
          'enabled': True,
          'enabled_supported': True,
          'ready_start_time': None,
          'status': 'StandBy',
          'target_level': 'Level3',
          'target_level_supported': True,
        }),
      }),
      dict({
        'code': 'CMPreExtraction',
        'index': 1,
        'output': dict({
          'available_modes': list([
            'PreBrewing',
            'PreInfusion',
            'Disabled',
          ]),
          'mode': 'PreInfusion',
          'times': dict({
            'time_in': dict({
              'seconds': 0.0,
              'seconds_max': dict({
                'pre_brewing': 9.0,
                'pre_infusion': 9.0,
              }),
              'seconds_min': dict({
                'pre_brewing': 2.0,
                'pre_infusion': 2.0,
              }),
              'seconds_step': dict({
                'pre_brewing': 0.1,
                'pre_infusion': 0.1,
              }),
            }),
            'time_out': dict({
              'seconds': 4.0,
              'seconds_max': dict({
                'pre_brewing': 9.0,
                'pre_infusion': 25.0,
              }),
              'seconds_min': dict({
                'pre_brewing': 1.0,
                'pre_infusion': 1.0,
              }),
              'seconds_step': dict({
                'pre_brewing': 0.1,
                'pre_infusion': 0.1,
              }),
            }),
          }),
        }),
      }),
      dict({
        'code': 'CMPreBrewing',
        'index': 1,
        'output': dict({
          'available_modes': list([
            'PreBrewing',
            'PreInfusion',
            'Disabled',
          ]),
          'dose_index_supported': False,
          'mode': 'PreInfusion',
          'times': dict({
            'pre_brewing': list([
              dict({
                'dose_index': 'ByGroup',
                'seconds': dict({
                  'seconds_in': 5.0,
                  'seconds_out': 5.0,
                }),
                'seconds_max': dict({
                  'seconds_in': 9.0,
                  'seconds_out': 9.0,
                }),
                'seconds_min': dict({
                  'seconds_in': 1.0,
                  'seconds_out': 1.0,
                }),
                'seconds_step': dict({
                  'seconds_in': 0.1,
                  'seconds_out': 0.1,
                }),
              }),
            ]),
            'pre_infusion': list([
              dict({
                'dose_index': 'ByGroup',
                'seconds': dict({
                  'seconds_in': 0.0,
                  'seconds_out': 4.0,
                }),
                'seconds_max': dict({
                  'seconds_in': 25.0,
                  'seconds_out': 25.0,
                }),
                'seconds_min': dict({
                  'seconds_in': 1.0,
                  'seconds_out': 1.0,
                }),
                'seconds_step': dict({
                  'seconds_in': 0.1,
                  'seconds_out': 0.1,
                }),
              }),
            ]),
          }),
        }),
      }),
      dict({
        'code': 'CMBackFlush',
        'index': 1,
        'output': dict({
          'last_cleaning_start_time': None,
          'status': 'Off',
        }),
      }),
    ]),
  })
# ---
